File: code/design-patterns/creational/object-pool/go/object_pool.go
package main

import (
	"fmt"
	"sync"
)

//////////////////////////////////// Bullet ////////////////////////////////////

type Bullet struct {
	Position int
	Active   bool
}

func NewBullet() *Bullet {
	return &Bullet{Position: 0, Active: false}
}

func (b *Bullet) Move() {
	b.Position += 1 // Simulates the bullet moving forward
}

func (b *Bullet) Draw() {
	fmt.Printf("Drawing bullet at position %d\n", b.Position)
}

//////////////////////////////////// BulletPool ////////////////////////////////////

type BulletPool struct {
	pool []*Bullet
	mu   sync.Mutex
}

var instance *BulletPool
var once sync.Once

// GetInstance returns the singleton instance of BulletPool
func GetInstance() *BulletPool {
	once.Do(func() {
		instance = &BulletPool{
			pool: make([]*Bullet, 0, 10), // Initialize with capacity of 10
		}
	})
	return instance
}

// Acquire retrieves a Bullet from the pool or creates a new one if none are available
func (p *BulletPool) Acquire() *Bullet {
	p.mu.Lock()
	defer p.mu.Unlock()
	if len(p.pool) > 0 {
		bullet := p.pool[len(p.pool)-1]
		p.pool = p.pool[:len(p.pool)-1]
		bullet.Active = true
		return bullet
	}
	return NewBullet() // Create a new bullet if the pool is empty
}

// Release adds a Bullet back to the pool if there is capacity
func (p *BulletPool) Release(bullet *Bullet) {
	p.mu.Lock()
	defer p.mu.Unlock()
	if len(p.pool) < cap(p.pool) {
		bullet.Active = false
		bullet.Position = 0 // Reset position
		p.pool = append(p.pool, bullet)
	}
}

File: code/design-patterns/python/creational/object-pool/object_pool.py
class Bullet:
    def init(self):
        self.position = None

    def move(self):
        # Move the bullet on the screen
        pass

    def draw(self):
        # Draw the bullet on the screen
        pass


class BulletPool:
    _instance = None
    maxBullets = 10  # Setting the maximum bullets allowed in the pool

    @staticmethod
    def getInstance():
        if BulletPool._instance is None:
            BulletPool._instance = BulletPool(BulletPool.maxBullets)
        return BulletPool._instance

    def __init__(self, size):
        if BulletPool._instance is not None:
            raise Exception("This class is a singleton!")
        else:
            # Initialize the pool with up to 'size' Bullets, capped by maxBullets
            self.pool = [Bullet() for _ in range(min(size, BulletPool.maxBullets))]

    def acquire(self):
        if self.pool:
            return self.pool.pop()
        else:
            return Bullet()  # Optionally, you might want to handle this differently

    def release(self, bullet: Bullet):
        if len(self.pool) < BulletPool.maxBullets:
            self.pool.append(bullet)
        else:
            # Optionally handle or log when the pool is at capacity and cannot accept more bullets
            pass
